ghcide/test/exe/FindImplementationAndHoverTests.hs

```diff
@@ -0,0 +1,269 @@
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+module FindImplementationAndHoverTests (tests) where
+
+import           Control.Monad
+import           Data.Foldable
+import           Data.Maybe
+import qualified Data.Text                  as T
+import qualified Language.LSP.Protocol.Lens as L
+import           Language.LSP.Test
+import           Text.Regex.TDFA            ((=~))
+
+import           Config
+import           Control.Category           ((>>>))
+import           Control.Lens               ((^.))
+import           Data.Text                  (Text)
+import           Development.IDE.Test       (standardizeQuotes)
+import           Test.Hls
+import           Test.Hls.FileSystem        (copyDir)
+
+tests :: TestTree
+tests = let
+  tst :: (TextDocumentIdentifier -> Position -> Session a, a -> Session [Expect] -> Session ()) -> Position -> String -> Session [Expect] -> String -> TestTree
+  tst (get, check) pos sfp targetRange title =
+    testWithDummyPlugin title (mkIdeTestFs [copyDir "hover"]) $ do
+        doc <- openDoc sfp "haskell"
+        waitForProgressDone
+        _x <- waitForTypecheck doc
+        found <- get doc pos
+        check found targetRange
+
+  checkHover :: (HasCallStack) => Maybe Hover -> Session [Expect] -> Session ()
+  checkHover hover expectations = traverse_ check =<< expectations where
+
+    check :: (HasCallStack) => Expect -> Session ()
+    check expected =
+      case hover of
+        Nothing -> unless (expected == ExpectNoHover) $ liftIO $ assertFailure "no hover found"
+        Just Hover{_contents = (InL MarkupContent{_value = standardizeQuotes -> msg})
+                  ,_range    = rangeInHover } ->
+          case expected of
+            ExpectRange  expectedRange -> checkHoverRange expectedRange rangeInHover msg
+            ExpectHoverRange expectedRange -> checkHoverRange expectedRange rangeInHover msg
+            ExpectHoverText snippets -> liftIO $ traverse_ (`assertFoundIn` msg) snippets
+            ExpectHoverExcludeText snippets -> liftIO $ traverse_ (`assertNotFoundIn` msg) snippets
+            ExpectHoverTextRegex re -> liftIO $ assertBool ("Regex not found in " <> T.unpack msg) (msg =~ re :: Bool)
+            ExpectNoHover -> liftIO $ assertFailure $ "Expected no hover but got " <> show hover
+            _ -> pure () -- all other expectations not relevant to hover
+        _ -> liftIO $ assertFailure $ "test not expecting this kind of hover info" <> show hover
+
+  extractLineColFromHoverMsg :: T.Text -> [T.Text]
```

Comment:
dupe

Reply:
This will likely change in the future, so I would like to keep it a duplicate.